Incident Response Plan (#2679)
WordPress/openverse: documentation/meta/incidents/index.md

> # Responding to Incidents

Author: This document can be reviewed in full 😄

[Making WordPress Chat](https://make.wordpress.org/chat/), to the `#openverse`
channel.
```

This document defines the policy for handling production incidents that occur
within Openverse. The goal is to ensure that user-facing service disruptions are
handled promptly, efficiently, and effectively, mitigating impact on our users
and maintaining the highest level of service availability possible.

## What is an incident?

An **incident** is any event which compromises or could compromise the goals of
the Openverse project. While commonly technical problems with our services
(Openverse.org, the Openverse API, and "official" integrations like the WP core
integration and pattern directory integrations) these can also be incidents of a
public relations or community nature. Here are some example incidents:

- The API is serving slower-than-typical responses which occasionally time out
- Openverse.org has a critical bug preventing search for mobile users using RTL
  languages
- A user has posted public comments using hateful language towards an Openverse
  maintainer
- Openverse.org is completely offline

> - Production resource usage is skyrocketing and could lead to downtime, but
>   there aren’t any actual issues yet

Reviewer: ```suggestion
- Production resource usage is skyrocketing and could lead to downtime, but
  users are not yet impacted
```

Totally a nit, but I'd posit that skyrocketing resource usage is an "issue" :slightly_smiling_face:

All of these scenarios suggest that **Openverse behavior normally depended on or
expected by regular users is not behaving as it should.**

It is crucial to be proactive here; if something _might_ be an incident, but
you’re not certain, it is better to record it than to ignore it. Do not assume
"someone else" is handling an incident unless they have explicitly recorded it
using the process defined here.

## How do we _identify_ incidents?

Incidents can be detected through various channels such as automated alerts,
user reports, or through manual checks by maintainers. The Mini Support Rotation
(MSR) runner makes these checks as part of their daily responsibilities.

While the MSR runner should be monitoring our communication channels closely,
**any online maintainer who sees something that might be an incident should

> follow the incident recording process outlined here.**

Reviewer: Just a note here. Most weeks I am the first person who logs on after the longest period of time with no one regularly online. That means that _anything_ that happens over the weekend inevitably falls on my plate to report.

For critical issues, that's fine, we understand that they're disruptive. But if there are multiple incidents of varying severity over a weekend, it is incredibly tedious to start the week of writing back-to-back incident reports. I did that this week, in fact, following the ethic described in this section, but distinctly felt that it could have been fine record _one_ of them and then explicitly delegated recording the other one to someone else, whenever someone else logged in for the week.

Basically: I think it's worth having some explicit process for acknowledging an incident, adding a note that it isn't severe, and then delegating (probably to MSR, who can then delegate it to someone else if needed) the actual full initial incident report.

Incident reports are very tedious and important to write with care. The tediousness of writing even one incident report (which I'm not complaining about, it's an unavoidable fact) means the requisite amount of care is less likely to always be applied when writing the second (or god forbid third, fourth, etc) report immediately after the previous one.

This is less about a broad equitability of incident report management and more just asking for a process to follow that can respect this ethic without requiring a full incident report (and still making clear that someone needs to do that bit).

Reviewer 2: Definitely support this thought, especially given we've run into the exact scenario you describe (once if not several times) before!

## How do we _classify_ incidents?

Incidents are classified in two ways:

- **Severity** - How impactful the incident is
- **Status** - The current state of the incident

### Severity

> Incidents should be classified using one of three levels of severity. It is
> important to note that regardless of severity, working to resolve incidents
> always **takes priority over regular project work and MSR duties.** Once service
> is stabilized, any ongoing investigations, preventative measures, or
> documentation changes will be prioritized by the team alongside our other work.

Reviewer: ```suggestion
Incidents should be classified using one of three levels of severity. It is
important to note that regardless of severity, working to stabilize services (or whatever harm mitigation is relevant for the incident)
always **takes priority over regular project work and MSR duties.** Once service
is stabilized, any ongoing investigations, preventative measures, or
documentation changes will be prioritized by the team alongside our other work.
```

Just two clarifications here. Incident "resolution" is a little ambiguous, but it also seems worthwhile to note that it isn't just service stability that takes precedence with respect to incidents. In the example incident you shared at the start of the document of hateful conduct, there could be important harm mitigation and care needed that should also take precedence.

Reviewer 2: +1 to this, "stabilize" seems like the more important thing to prioritize.

This severity scale is intentionally distinct from our GitHub priority scale.
Any GitHub issues created to _stabilize_ a Severity 1 or Severity 2 incident
should be marked as "critical" in GitHub. Other issues should be prioritized at
your own discretion.

- **Severity 1**: Critical issue affecting all users and all functionalities of

>   Openverse.

Reviewer: What does "all functionalities of Openverse" entail? I understand there's some ambiguity here that is necessary and folks need to be comfortable making a decision on the fly, but severity 1 would only happen if, for example, AWS `us-east-1` lost all availability zones. Which, to be clear, is something we can basically do nothing about in the moment.

I guess, asked another way, can you share examples of what would qualify for severity 1? A full frontend outage would be, to me, as severe as we should ever get (within our control), but would just be severity 2 in this scale (API and its users are okay).

- **Severity 2**: Major issue affecting a large number of users or critical
  functionality of Openverse.
- **Severity 3**: Minor issue affecting a limited number of users or
  non-critical functionality of Openverse.

### Status

- **Stabilization pending** - The incident is actively disrupting service
- **Stabilized** - The incident is no longer directly disrupting service. root
  causes may
- **Under investigation** - The root causes of the incident are being
  researched. Stabilized but not yet resolved
- **Resolved** - The incident is stable, and long-term fixes to prevent future
  occurrences or other mitigations have been identified.

As you might expect, incidents can be _destabilized_ if they reoccur, and the
state should rollback to "Stabilization pending":

```{mermaid}
graph LR
    A[Stabilization pending]
    B[Stabilized]
    C[Under investigation]
    D[Resolved]

    A --> B
    B --> C
    B --> A
    C --> A
    C --> D
    D --> A
```

## How do we _handle_ incidents?

```{warning}
Dealing with an incident now? Go to the [Incident Response Runbook](/meta/incidents/runbooks/incident_response_runbook).

> ```

Reviewer 2: I like this addition here 🙂

This section addresses the philosophy of our approach to incident management and
explanation behind the steps in the
[runbook](/meta/incidents/runbooks/incident_response_runbook).

There are four main steps to handling production incidents:

<!-- prettier-ignore -->
| # | Name | Goal | Status After |
|---|------|------|--------------|
| 1 | Record | Visibly acknowledge the incident and create the communication threads necessary for ongoing discussion and investigation. Understand how bad the situation is based on our severity definitions. | Stabilization pending OR Stabilized (as appropriate)
| 2 | Stabilize |<p>If there is a total or partial outage, identify service stabilization options and action them. | Stabilized |
| 3 | Resolve | Identify outstanding improvements to the service that would mitigate the issue in the future and create tickets to implement. | Under investigation |
| 4 | Review | Identify process improvements. If the underlying cause is still unknown, identify further steps to resolve or continue investigation of the issue via 5-whys. | Under investigation OR Resolved (as appropriate) |

> All maintainers are required to report incidents they encounter. **Crucially,
> maintainers who record incidents are not _inherently_ responsible for
> stabilizing them.** Their only required job is to record the incident. Every
> incident has a **Lead** who is the person responsible for providing updates on
> the incident and delegating tasks. The Lead is also responsible for hosting the
> retrospective after the incident is resolved.

Reviewer: This is a very helpful distinction and mostly covers the concern I raised above about needing to write multiple incident reports in a row for e.g., severity 3 things.

I'll suggest one small change for clarity:

```suggestion
All maintainers are required to record incidents they encounter. **Crucially,
maintainers who record incidents are not _inherently_ responsible for
stabilizing them.** Their only required job is to record the incident. Every
incident has a **Lead** who is the person responsible for providing updates on
the incident and delegating tasks. The Lead is also responsible for hosting the
retrospective after the incident is resolved.
```

Using "record" as the verb rather than "report" when we mean "the first and only necessary immediate step for all incidents regardless of severity" would help with consistency and clarity of the precise expectation.

### How do we determine the lead?

By default, this is the week’s MSR runner. If it is outside of the MSR runner’s
working hours or they are unresponsive (with later explanation, ideally) someone
else will need to be the lead. It is the recorder’s job to find this lead or act
as this lead. Factors which should be considered in finding a lead are
availability and expertise. Is anyone online uniquely qualified to take this
work on? Alternatively, has one person been handling most incidents and this
time, someone new should take on? Consulting the threads for past incidents will
make it easier to see the distribution of incident leadership so we can easily

> find who hasn’t done it recently.

Reviewer: Over time we'll come to understand this better, but a useful clarification here could be that stabilisation of >= severity 2 incidents _probably_ takes priority over finding a lead, if the de facto lead is unavailable. Mostly, I want to avoid the team or any individual on the team to accidentally think they both need to try to work on stabilisation _and_ find a lead (e.g., that they are simultaneous priorities and expectations) _or_ risk the assumption that the person leading stabilisation for critical issues is also necessarily the lead.

# Responding to Incidents

```{note}
Currently experiencing an issue with Openverse? For maintainers, please jump to the [Incident Response Runbook](/meta/incidents/runbooks/incident_response_runbook). Otherwise,
please [file a bug report](https://github.com/WordPress/openverse/issues/new?template=bug_report.md&title=Production+Incident:+Describe+the+incident) in GitHub or send a message in the
[Making WordPress Chat](https://make.wordpress.org/chat/), to the `#openverse`

> channel.

Reviewer: For clarity, we should link to the security disclosure form here in addition to the bug report template. https://github.com/WordPress/openverse/security/advisories/new

using the process defined here.

## How do we _identify_ incidents?

Incidents can be detected through various channels such as automated alerts,
user reports, or through manual checks by maintainers. The Mini Support Rotation
(MSR) runner makes these checks as part of their daily responsibilities.

While the MSR runner should be monitoring our communication channels closely,
**any online maintainer who sees something that might be an incident should
follow the incident recording process outlined here.** This does not mean that
the reporting maintainer is responsible for stabilizing the incident or acting
as the incident lead. It simply means that they are responsible for making the
initial report on the incident.

## How do we _classify_ incidents?

Incidents are classified in two ways:

- **Severity** - How impactful the incident is
- **Status** - The current state of the incident

### Severity

Incidents should be classified using one of three levels of severity. It is
important to note that regardless of severity, working to stabilize services (or
whatever harm mitigation is relevant for the incident) always **takes priority

> over regular project work and MSR duties.** Once service is stabilized, any

Reviewer 3: +1 to @sarayourfriend suggestion of pulling out of the scope the non-technical issues. It doesn't seems like they should follow the same process.

> over regular project work and MSR duties.** Once service is stabilized, any

Reviewer 4: +1 as well, especially given @sarayourfriend 's examples -- making this distinction and limiting the scope saves a lot of questions/clarifications.

ongoing investigations, preventative measures, or documentation changes will be
prioritized by the team alongside our other work.

This severity scale is intentionally distinct from our GitHub priority scale.
Any GitHub issues created to _stabilize_ a Severity 1 or Severity 2 incident
should be marked as "critical" in GitHub. Other issues should be prioritized at
your own discretion.

- **Severity 1**: Critical issue affecting all/most users and all/most
  functionalities of Openverse.
- **Severity 2**: Major issue affecting a large number of users or critical
  functionality of Openverse.
- **Severity 3**: Minor issue affecting a limited number of users or
  non-critical functionality of Openverse.

### Status

- **Stabilization pending** - The incident is actively disrupting service

> - **Stabilized** - The incident is no longer directly disrupting service. root
>   causes may

Reviewer 3: Looks like the sentence is incomplete here, maybe something like this was what followed (?)
```suggestion
- **Stabilized** - The incident is no longer directly disrupting service. Root
  causes may still be unknown.
```

Reviewer 4: That would seem to overlap with `Under investigation` 🤔 I'm actually not sure what the difference between the two should be?

Reviewer 2: I tried to add some clarifying language around that transition Staci, hopefully that helps!

```

This section addresses the philosophy of our approach to incident management and
explanation behind the steps in the
[runbook](/meta/incidents/runbooks/incident_response_runbook).

There are four main steps to handling production incidents:

<!-- prettier-ignore -->
| # | Name      | Goal                                                                                                                                                                                             | Status After                                         | Tag                  |
|---|-----------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------|----------------------|
| 1 | Record    | Visibly acknowledge the incident and create the communication threads necessary for ongoing discussion and investigation. Understand how bad the situation is based on our severity definitions. | Stabilization pending OR Stabilized (as appropriate) | `#status-recorded`   |

> | 2 | Stabilize | <p>If there is a total or partial outage, identify service stabilization options and action them.                                                                                                | Stabilized                                           | `#status-stabilized` |

Reviewer: Stray HTML tag.

```suggestion
| 2 | Stabilize | If there is a total or partial outage, identify service stabilization options and action them.                                                                                                | Stabilized                                           | `#status-stabilized` |
```

| 3 | Resolve   | Identify outstanding improvements to the service that would mitigate the issue in the future and create tickets to implement.                                                                    | Under investigation                                  | `#status-resolved`   |
| 4 | Review    | Identify process improvements. If the underlying cause is still unknown, identify further steps to resolve or continue investigation of the issue via 5-whys.                                    | Under investigation OR Resolved (as appropriate)     | `#status-reviewed`   |

All maintainers are required to record incidents they encounter. **Crucially,
maintainers who record incidents are not _inherently_ responsible for
stabilizing them.** Their only required job is to record the incident. Every
incident has a **Lead** who is the person responsible for providing updates on
the incident and delegating tasks. The Lead is also responsible for hosting the

> retrospective after the incident is resolved.

Reviewer: This qualification is repeated from above. We should qualify this once and if necessary refer or link to the specific section later, rather than re-stating it several times.

### How do we determine the lead?

> By default, this is the week’s MSR runner. If it is outside the MSR runner’s
> working hours, or they are unresponsive (with later explanation, ideally)
> someone else will need to be the lead. It is the recorder’s job to find this
> lead or act as this lead; although for incidents that have a severity <= 2,
> stabilizing the service may be the highest priority and take precedence over
> finding a lead. Factors which should be considered in finding a lead are
> availability and expertise. Is anyone online uniquely qualified to take this
> work on? Alternatively, has one person been handling most incidents and this
> time, someone new should take on? Consulting the threads for past incidents will
> make it easier to see the distribution of incident leadership so we can easily
> find who hasn’t done it recently.

Reviewer: Rewording some things here:
* Remove language that can be easily misinterpreted as accusatory
* Remove the "less than or equal to severity 2". It is confusing to think of severity levels in terms of their mathematical order (maybe we should switch to colours or something like "high"/"medium"/"low"), especially because they are in the opposite order of "quantity" of severity. Severity 1 is _more_ severe than severity 2, but is "less than" severity 2 in the mathematical sense the original version used. It's clearer to write out the levels we mean individually.
* Simplify the lead 'considerations' sentences.

```suggestion
The current MSR runner is the default incident lead. If it is outside the MSR runner’s
working hours, or they are unable to lead the incident response, then
someone else will need to be the lead. It is the recorder’s job to find this
lead or act as this lead; for incidents that are severity 1 or 2,
stabilizing the service may be the highest priority and take precedence over
finding a lead. Factors which should be considered in finding a lead are:
* Availability
  * Has one person been handling most incidents and this time, someone new should take on?
  * Consult threads for past incidents to see the distribution of recent incident leadership and find someone who hasn’t recently lead an incident.
* Expertise. Is anyone online uniquely qualified to guide the resolution?
```

> We can use our discretion in making these decisions, and if we find that the
> distribution of Lead responsibilities is inequitably distributed we can consider
> formal processes for finding them.

Reviewer: suggestion: remove the sense that we might _not_ use our discretion. We will always use our discretion for this, it's impossible not to. Additionally, remove the unintentional implication that this process is somehow informal (could read "unintentional") or that a "formal" process is inherently more equitable.

```suggestion
This process is intentionally flexible to allow for the rapidly evolving nature of incidents and to avoid too much rigidity. If the distribution of lead responsibilities is inequitably distributed, we will develop alternative processes for assigning leads.
```

# Responding to Incidents

```{note}
Currently experiencing an issue with Openverse? For maintainers, please jump to the [Incident Response Runbook](/meta/incidents/runbooks/incident_response_runbook). Otherwise,
please [file a bug report](https://github.com/WordPress/openverse/issues/new?template=bug_report.md&title=Production+Incident:+Describe+the+incident) in GitHub,
[submit a security advisory](https://github.com/WordPress/openverse/security/advisories/new), or send a message in the
[Making WordPress Chat](https://make.wordpress.org/chat/), to the `#openverse`
channel.
```

This document defines the policy for handling technical production incidents
that occur within Openverse. The goal is to ensure that user-facing service
disruptions are handled promptly, efficiently, and effectively, mitigating
impact on our users and maintaining the highest level of service availability
possible.

## What is an incident?

> An **incident** in the context of these documents is any event which affects the
> ability of Openverse to behave normally in a way that is depended on or expected
> by regular users. This document and the referenced runbooks refer to incidents

Reviewer: nit: This is kind of a wordy/confusing way to describe an incident. I think I see why, it's trying to get at a broad definition of technical issues, but feels like we can simplify the sentence by reducing the negations and qualifications without diluting the broadness of what we consider a technical "incident".

```suggestion
An **incident** in the context of these documents is anything that causes Openverse
to behave differently than the expected or depended behavior.
This document and the referenced runbooks refer to incidents
```

Although, I think we could get rid of the first sentence altogether. The rest of the section does a great job of describing what an incident is without the need to stuff a summary into the introductory sentence. This is my preferred suggestion (if you make a modification at all). This is not a blocker.

```suggestion
This document and the referenced runbooks refer to incidents
```

Reviewer 2: Thanks for the suggestion, removed in 69ca231652cfa9847fb8ab8b76d23db120cc8187!

as technical problems with our services (Openverse.org, the Openverse API, and
"official" integrations like the WP core integration and pattern directory
integrations). Here are some example incidents:

- The API is serving slower-than-typical responses which occasionally time out
- Openverse.org has a critical bug preventing search for mobile users using RTL

>   languages maintainer

Reviewer: ```suggestion
  languages
```

to note that regardless of severity, working to stabilize services (or whatever
harm mitigation is relevant for the incident) always **takes priority over
regular project work and MSR duties.** Once services are stabilized, any ongoing
investigations, preventative measures, or documentation changes will be
prioritized by the team alongside our other work.

This severity scale is intentionally distinct from our GitHub priority scale.
Any GitHub issues created to _stabilize_ a Severity 1 or Severity 2 incident are
marked as "critical" in GitHub. Other issues are prioritized at your own
discretion.

- **Severity 1**: Critical issue affecting all/most users and all/most
  functionalities of Openverse.
- **Severity 2**: Major issue affecting a large number of users or critical
  functionality of Openverse.
- **Severity 3**: Minor issue affecting a limited number of users or
  non-critical functionality of Openverse.

### Status

- **Stabilization pending** - The incident is actively disrupting service.
- **Stabilized** - The incident is no longer directly disrupting service. Root
  causes may still be unknown.
- **Under investigation** - The root causes of the incident are being

>   researched. Stabilized but not yet resolved.

Reviewer: What's the difference between stabilized and under investigation? Is that to allow specific time after stabilisation where we haven't started looking into the root cause?

Reviewer 4: I also had this question. Per the new text added below:

> Some incidents may not require further investigation once they are stabilized and can be immediately resolved.

But I think those incidents would go `Stabilization pending` straight to `Resolved`, so I'm still not sure when something would be `Stabilized` but not `Under investigation`. I guess it implies that investigation is needed but not started? Or, possibly, investigation has already happened and the cause is identified but not yet fixed? Although that can't be right, because that's the definition of `Resolved` given below.

Reviewer 4: Just noticing that you commented on that for Resolved as well @sarayourfriend  [here](https://github.com/WordPress/openverse/pull/2679/files#r1421862675). _If_ Resolved is updated to mean that the long-term fixes are _implemented_, then one interpretation of the statuses which would maybe make sense  is:

* Stabilization pending: actively disrupting serice
* Under investigation: stabilized but not resolved. Root causes are being researched.
* Stabilized: Root causes identified but not yet fixed
* Resolved: Long-term fixes applied

But then the name `stabilized` feels like an inadequate description 😓

Reviewer 2: Thanks for the further discussion here! I was trying to identify a state where things are no longer on fire, but perhaps we're not sure if things _need_ investigation or if they can be resolved. I honestly hadn't thought of moving straight from `Stabilization pending` to `Resolved` 😮 Perhaps that would cover the case I'm imagining well enough? Regardless, I'll make an issue for clarifying this!

Reviewer 2: Issue created: https://github.com/WordPress/openverse/issues/3579

- **Resolved** - The incident is stable, and long-term fixes to prevent future

>   occurrences or other mitigations have been identified.

Reviewer: Can you clarify if the resolution of an incident only involves _identifying_ mitigations rather than implementing them? So far we've only resolved incidents after we've implemented the fixes, I think, rather than just identifying them. My perception of that could be wrong though, just thinking of the API response time incident and the one this incident response plan grew out of (https://github.com/WordPress/openverse/issues/2060).

I'm not necessarily objecting to that being the case, I just wanted to clarify if that's an intentional difference from the ad-hoc process we've followed so far. This also seems to contradict the later section about process rationale which says that resolved is "Complete the immediately necessary work to prevent this issue from recurring in the future.".

Though I think maybe my confusion is coming from the past tense of this status "resolved" rather than "resolve". Is there an intermediate status, "resolving", before "resolved"? I might also be mixing up the difference between the incident status and the status of work related to the incident. Perhaps they're separate and maybe "resolved" isn't the best word. Maybe something like "investigated" instead?

I see this as potentially similar to the distinction between "shipped" and "success" in our project statuses; or just a misunderstanding on my part of how the work to implement mitigations for an incident is related to the incident status.

Reviewer 2: Oh thank you for pointing this out - I think I missed this in the changing of the states! I think `Resolved` should represent that the mitigations/fixes have been _implemented_ to prevent this from coming up in the future, but there may still be extant issues that would assist future disruptions (like new logging/metrics capture). I'll reword this hear to clarify a bit.

- **Reviewed** - A retrospective has been completed for the incident.

As you might expect, incidents can be _destabilized_ if they reoccur, and the
state should roll back to "Stabilization pending". Some incidents may not
require further investigation once they are stabilized and can be immediately

> resolved.

Reviewer: What's the rationale for _destabilising_ an incident rather than treating it as a wholly new incident (at least initially)? It feels like a big decision to make during an incident ("is this spike in average response times caused by the same thing as last week or is it a new thing altogether?").

Reviewer 2: I think an issue would be destabilized if the exact same behavior that caused the initial incident was exhibited after a fix (which was presumed to address the issue) was deployed. It seems more probable to me that a fix might not address the ongoing incident than that we would have two, separate incidents occur at the same time.

That said, I think there's some temporality to destabilization. For instance, if we're more than 48 hours from resolving an incident, a new incident should probably be created. I'll add some language to this extent, let me know if you feel there's still ambiguity after and I'll make an issue to clarify that more.
